eventsources/sources/bitbucketserver/start.go

```diff
@@ -190,26 +192,34 @@ func (el *EventListener) StartListening(ctx context.Context, dispatch func([]byt
 	bitbucketConfig.AddDefaultHeader("x-atlassian-token", "no-check")
 	bitbucketConfig.AddDefaultHeader("x-requested-with", "XMLHttpRequest")
 
-	// When running multiple replicas of the eventsource, they will all try to create the webhook.
-	// Randomly sleep some time to mitigate the issue.
-	s1 := rand.NewSource(time.Now().UnixNano())
-	r1 := rand.New(s1)
-	time.Sleep(time.Duration(r1.Intn(2000)) * time.Millisecond)
-
 	ctx, cancel := context.WithCancel(ctx)
 	defer cancel()
 
 	ctx = context.WithValue(ctx, bitbucketv1.ContextAccessToken, bitbucketToken)
-	err = router.CreateBitbucketWebhook(ctx, bitbucketConfig)
 
-	if err != nil {
-		logger.Errorw("failed to create/update Bitbucket webhook", zap.Error(err))
+	createWebhooks := func() {
+		for _, repo := range bitbucketserverEventSource.GetBitbucketServerRepositories() {
+			if err = router.CreateBitbucketWebhook(ctx, bitbucketConfig, repo); err != nil {
+				logger.Errorw("failed to create/update Bitbucket webhook",
+					zap.String("project-key", repo.ProjectKey), zap.String("repository-slug", repo.RepositorySlug), zap.Error(err))
+				continue
+			}
+
+			time.Sleep(500 * time.Millisecond)
+		}
 	}
 
+	// When running multiple replicas of the eventsource, they will all try to create the webhook.
+	// Randomly sleep some time to mitigate the issue.
+	s1 := rand.NewSource(time.Now().UnixNano())
+	r1 := rand.New(s1)
+	time.Sleep(time.Duration(r1.Intn(2000)) * time.Millisecond)
+	createWebhooks()
+
 	go func() {
 		// Another kind of race conditions might happen when pods do rolling upgrade - new pod starts
 		// and old pod terminates, if DeleteHookOnFinish is true, the hook will be deleted from Bitbucket.
-		// This is a workround to mitigate the race conditions.
+		// This is a workaround to mitigate the race conditions.
```

Comment:
Similar workaround has been existing in several event sources, do you think it would be better if we let the controller to create a K8s job to create the hook when there's a new or updated event source object, and another job to delete the hook when the event source object is deleted?

Reply:
Thats an interesting idea actually, but what if the user decides to update the webhook configuration in the event source object? Wouldn't the controller need to create a Job for deleting the old hook as well? Because if so, I think we would still face race conditions with this solution as well

Reply:
I think the biggest issue for updating is, we don't know what is the old hook, because the controller only knows the updated object.....

Reply:
Hmm yes thats a good point. I mean we can always store some state in the event source objects (etcd) and then use it on update events by reading it from the old object state, but this would definitely introduce a whole lot of new complexity to the system..

Reply:
No.. When an updating happens, the event source object is updated in etcd before the controller sees it.

Reply:
But wouldn't the controller be able to see the old object as well as the new object on update events?

Reply:
It is swallowed by controller-runtime. https://github.com/kubernetes-sigs/controller-runtime/issues/399